dart/common/Addon.h

```diff
@@ -0,0 +1,305 @@
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_ADDON_H_
+#define DART_COMMON_ADDON_H_
+
+#include <string>
+
+#include "dart/common/Extensible.h"
+#include "dart/common/detail/NoOp.h"
+
+namespace dart {
+namespace common {
+
+class AddonManager;
+
+class Addon
+{
+public:
+
+  friend class AddonManager;
+
+  /// If your Addon has a State class, then that State class should inherit this
+  /// Addon::State class. This allows us to safely serialize, store, and clone
+  /// the states of arbitrary Addon extensions. If your Addon is stateless, then
+  /// you do not have to worry about extending this class, because
+  /// Addon::getState() will simply return a nullptr by default, which is taken
+  /// to indicate that it is stateless.
+  ///
+  /// The distinction between the State class and the Properties class is that
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class State : public Extensible<State> { };
+
+  /// Use the StateMixer class to easily create a State extension from an
+  /// existing class or struct.
+  template <class Mixin>
+  using StateMixer = ExtensibleMixer<State, Mixin>;
+
+  /// If your Addon has a Properties class, then it should inherit this
+  /// Addon::Properties class. This allows us to safely serialize, store, and
+  /// clone the properties of arbitrary Addon extensions. If your Addon has no
+  /// properties, then you do not have to worry about extending this class,
+  /// because Addon::getProperties() will simply return a nullptr by default,
+  /// which is taken to indicate that it has no properties.
+  ///
+  /// The distinction between the State class and the Properties class is that
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class Properties : public Extensible<Properties> { };
+
+  /// Use the PropertiesMixer class to easily create a Properties extension
+  /// from an existing class or struct.
+  template <class Mixin>
+  using PropertiesMixer = ExtensibleMixer<Properties, Mixin>;
+
+  /// Virtual destructor
+  virtual ~Addon() = default;
+
+  /// Clone this Addon into a new manager
+  virtual std::unique_ptr<Addon> cloneAddon(AddonManager* newManager) const = 0;
+
+  /// Set the State of this Addon. By default, this does nothing.
+  virtual void setAddonState(const std::unique_ptr<State>& otherState);
+
+  /// Get the State of this Addon. By default, this returns a nullptr which
+  /// implies that the Addon is stateless.
+  virtual const State* getAddonState() const;
```

Comment:
The point of this function is to allow you to peek at the state data structure without needing to allocate new memory to do so.

When you call ``clone()`` on ``Addon::State`` like ``addon->getAddonState()->clone()`` it will return a ``std::unique_ptr<Addon::State>``, and that is what makes it symmetric with ``setAddonState(std::unique_ptr)``.